walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,65 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        if (isMeetingRequestTooLong(request)) {
+            return Collections.emptyList();
+        }
+
+        Collection<TimeRange> availableTimeRange = new ArrayList<TimeRange>();
+        availableTimeRange.add(TimeRange.WHOLE_DAY);
+
+        Collection<String> meetingAttendees = request.getAttendees();
+        for (Event e : events) {
+            Collection<String> eventAttendees = e.getAttendees();
+            if (hasCommonAttendees (meetingAttendees, eventAttendees)) {
+                TimeRange eventTime = e.getWhen();
+                updateAvailableTimeRange(availableTimeRange, eventTime, request.getDuration());                
+            }
+        }
+
+        return availableTimeRange;
+    }
+
+    private boolean hasCommonAttendees(Collection<String> meetingAttendees, Collection<String> eventAttendees) {
+        return !Collections.disjoint(meetingAttendees, eventAttendees);
+    }
+
+    private boolean isMeetingRequestTooLong(MeetingRequest request) {
+        return request.getDuration() > TimeRange.WHOLE_DAY.duration();
+    }
+
+    private boolean isTimeslotLongEnough(TimeRange timeslot, long meetingDuration) {
+        return timeslot.duration() >= meetingDuration;
+    }
+
+    private void updateAvailableTimeRange(Collection<TimeRange> availableTimeRange, TimeRange eventTime, long meetingDuration) {
+        for (TimeRange t: new ArrayList<TimeRange>(availableTimeRange)) {
```

Comment:
Alternatively could use `Collections.unmodifiableList(availableTimeRange)`, if you pass in the `availableTimeRange` as `List` instead of `Collection`

Reply:
I actually tried Collections.unmodifiableCollection before and this led to a ConcurrentModificationException error, since I am modifying availableTimeRange while I am looping over it at the same time, hence I created a separate copy of the list to loop over, while modifying the original copy. This same error occurs with unmodifiableList.

Is there a better way you would recommend for such cases, where I would like to modify the list while I am looping over it?

Reply:
I see. In that case, this approach should be fine.